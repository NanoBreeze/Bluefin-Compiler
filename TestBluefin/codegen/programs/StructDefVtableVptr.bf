/*
Verifies that vtable and vptrs are created correctly.
For vtable, order of the func ptrs matter
*/

struct Base {
	virtual void one() {}
	virtual int two(bool a) {
		return 3;
	}
};

struct Der extends Base {
	void one() override {}

	virtual bool three() {
		return true;
	}

	int two(bool a) override {
		return 4;
	}


	void notVirtualMethod() {}	// this won't be placed in a vtable
};

struct DerDer extends Der {
	int two(bool a) override {
		return 5;
	}
};