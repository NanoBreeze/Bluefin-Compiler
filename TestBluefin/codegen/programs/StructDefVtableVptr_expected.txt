; ModuleID = 'StructDefVtableVptr.bf'
source_filename = "StructDefVtableVptr.bf"

%Base = type { i32 (...)** }
%Der = type { %Base }
%DerDer = type { %Der, i1 }
%NonVirtualBase = type {}
%VirtualDer = type { i32 (...)**, %NonVirtualBase, i32 }

@_vtable_Base = global [2 x i8*] [i8* bitcast (void (%Base*)* @_Base_one to i8*), i8* bitcast (i32 (%Base*, i1)* @_Base_two to i8*)]
@_vtable_Der = global [3 x i8*] [i8* bitcast (void (%Der*)* @_Der_one to i8*), i8* bitcast (i32 (%Der*, i1)* @_Der_two to i8*), i8* bitcast (i1 (%Der*)* @_Der_three to i8*)]
@_vtable_DerDer = global [3 x i8*] [i8* bitcast (void (%Der*)* @_Der_one to i8*), i8* bitcast (i32 (%DerDer*, i1)* @_DerDer_two to i8*), i8* bitcast (i1 (%Der*)* @_Der_three to i8*)]
@_vtable_VirtualDer = global [1 x i8*] [i8* bitcast (void (%VirtualDer*)* @_VirtualDer_here to i8*)]

define void @_Base_one(%Base* %this) {
entry:
  ret void
}

define i32 @_Base_two(%Base* %this, i1 %a) {
entry:
  %alloctmp_a = alloca i1, align 1
  store i1 %a, i1* %alloctmp_a, align 1
  ret i32 3
}

define void @_ctor_Base(%Base* %this) {
entry:
  %memCast = bitcast %Base* %this to i32 (...)***
  store i32 (...)** bitcast ([2 x i8*]* @_vtable_Base to i32 (...)**), i32 (...)*** %memCast, align 8
  ret void
}

define void @_Der_one(%Der* %this) {
entry:
  ret void
}

define i1 @_Der_three(%Der* %this) {
entry:
  ret i1 true
}

define i32 @_Der_two(%Der* %this, i1 %a) {
entry:
  %alloctmp_a = alloca i1, align 1
  store i1 %a, i1* %alloctmp_a, align 1
  ret i32 4
}

define void @_Der_notVirtualMethod(%Der* %this) {
entry:
  ret void
}

define void @_ctor_Der(%Der* %this) {
entry:
  %memCast = bitcast %Der* %this to %Base*
  call void @_ctor_Base(%Base* %memCast)
  %memCast1 = bitcast %Der* %this to i32 (...)***
  store i32 (...)** bitcast ([3 x i8*]* @_vtable_Der to i32 (...)**), i32 (...)*** %memCast1, align 8
  ret void
}

define i32 @_DerDer_two(%DerDer* %this, i1 %a) {
entry:
  %alloctmp_a = alloca i1, align 1
  store i1 %a, i1* %alloctmp_a, align 1
  ret i32 5
}

define void @_ctor_DerDer(%DerDer* %this) {
entry:
  %memCast = bitcast %DerDer* %this to %Der*
  call void @_ctor_Der(%Der* %memCast)
  %memCast1 = bitcast %DerDer* %this to i32 (...)***
  store i32 (...)** bitcast ([3 x i8*]* @_vtable_DerDer to i32 (...)**), i32 (...)*** %memCast1, align 8
  ret void
}

define void @_NonVirtualBase_one(%NonVirtualBase* %this) {
entry:
  ret void
}

define void @_ctor_NonVirtualBase(%NonVirtualBase* %this) {
entry:
  ret void
}

define void @_VirtualDer_here(%VirtualDer* %this) {
entry:
  ret void
}

define void @_ctor_VirtualDer(%VirtualDer* %this) {
entry:
  %memCast = bitcast %VirtualDer* %this to %NonVirtualBase*
  call void @_ctor_NonVirtualBase(%NonVirtualBase* %memCast)
  %memCast1 = bitcast %VirtualDer* %this to i32 (...)***
  store i32 (...)** bitcast ([1 x i8*]* @_vtable_VirtualDer to i32 (...)**), i32 (...)*** %memCast1, align 8
  ret void
}
